xarray/core/dataset.py

```diff
@@ -2179,5 +2179,65 @@ def real(self):
     def imag(self):
         return self._unary_op(lambda x: x.imag, keep_attrs=True)(self)
 
+    def get_variables_by_attributes(self, **kwargs):
+        """Returns a ``Dataset`` with variables that match specific conditions.
+        Can pass in ``key=value ``or ``key=callable``.  Variables are returned
+        that contain all of the matches or callable returns True.  If using a
+        callable note that it should accept a single parameter only,
+        the attribute value.
+
```

Comment:
I'd like to see a more formal docstring here, more like: 

```Python

    def get_variables_by_attributes(self, **kwargs):
        """Summary...

        Parameters
        ----------
        **kwargs : key=value
            key : str
                Description...
            value : str or callable
                Description...
        Returns
        -------
        new : Dataset
       
        Examples
        --------
        ...
    """"
```
This is similar to what we did in `da.rolling`: https://github.com/pydata/xarray/blob/master/xarray/core/rolling.py#L36-L41

```